Initial implementation

ekohl opened the pull request.

ekohl committed:
Initial implementation

evgeni commented on client/psf.sh:
```diff
@@ -0,0 +1,60 @@
+#!/bin/bash -e
```
I once learned not to use `-e` in the shebang, but explicitly call `set -e`, so that whoever thinks it's smart to call `bash psf.sh` gets the same treatment.

evgeni commented on server/direct/puppetserver-foreman:
```diff
@@ -0,0 +1,300 @@
+
+
+def load_json_or_yaml(data):
+    try:
+        return json.loads(data)
+    except:
+        try:
+            return yaml.safe_load(data)
+        except:
+            raise ValueError('Invalid JSON/YAML')
```
Given any JSON document is valid YAML, and you depend on PyYAML anyways, just always parse with YAML?

ekohl replied:
YAML parsing does have odd quirks so in general I'd prefer to avoid. However, I don't have a good content detection. Perhaps I can check if it starts with `---` and directly move to YAML.

ekohl commented:
I also debated using some HTTP server but `http.server` only gained `ThreadingHTTPServer` in Python 3.7 and preferred to keep the dependencies to a minimum. Also, Ruby's Net::HTTP can't connect to a unix socket out of the box. There is https://github.com/puppetlabs/net_http_unix but it'd introduce an additional dependency.

evgeni commented:
The only OS that has no Python 3.7 by default, is EL7, which I don't particularly care about for new code.

ekohl commented:
Updated. This is now working end-to-end and I tested the README instructions. In case anyone wants to try it out without building the RPM, you can use https://ekohl.nl/psf-0.1.0-1.el8.noarch.rpm for convenience.

I haven't tried to see what the minimal permissions should be. I just tested with an admin account.

cc @ares

ekohl committed:
WIP

This rearranges the directory layout a bit and has various fixes.  Also
renames the server to psfd. It makes an attempt at supporting a Foreman
Proxy setup, but it needs a new systemd unit file and has been untested.

If the regular Foreman API would accept client certificates, this code
would be easier.

Also adds Python type annotations to some functions.

lzap commented:
I took a look on this and I believe you forgot to push a file. Current version does not make any sense - data is not being sent anywhere. Can you update the description with the overall design of this? I would like to first understand the final design of this.

Overall I feel like this is overengineered. Puppet server can simply spawn a process for ENC, that is pretty straightforward and relatively fast. What you are building here, according to my understanding, is far more complex setup that does the same. This must come with some clear benefits and these should be communicated.

Other observations:

* I do not understand how this is supposed to work: `puppet config set --section server reports psf` that does not appear to be a reporter that ships with puppet master. I guess this is some leftover.

* Delivery method (send data to Foreman via HTTPS) must be extendable, we are working on a new way of sending reports (and facts) and this must be an easy change. The format will also need to change, the new plugin now supports native report format (YAML). But this is completely missing from the last commit, so I can't see what is the current status.

* Avoid using Ruby's built-in Timer, it is extremely slow and clunky (spawns a thread), UNIX sockets do provide normal timeouts via SO_RCVTIMEO and SNDTIMEO from the OS its just a bit hard use in Ruby: `timeval = [1, 0].pack("l_2")` and then `setsockopts` for both values. Alternatively, use IO select which is the elegant way but little bit more lines to do it.

* There are some concerns about EL7, does this mean you are not building this with EL7 support in mind?

ekohl committed:
WIP

ekohl committed:
Switch to proxy mode by default
